src/sentry/replays/usecases/reader.py

```diff
@@ -0,0 +1,256 @@
+    replay_id: str,
+    segment_id: int,
+) -> RecordingSegmentStorageMeta | None:
+    """Return filestore metadata derived from our Postgres table."""
+    segment = ReplayRecordingSegment.objects.filter(
+        project_id=project_id,
+        replay_id=replay_id,
+        segment_id=segment_id,
+    ).first()
+    if segment is None:
+        return None
+
+    return RecordingSegmentStorageMeta(
+        project_id=project_id,
+        replay_id=replay_id,
+        segment_id=segment.segment_id,
+        retention_days=0,
+        date_added=segment.date_added,
+        file_id=segment.file_id,
+    )
+
+
+def fetch_direct_storage_segments_meta(
+    project_id: int,
+    replay_id: str,
+    offset: int,
+    limit: int,
+) -> List[RecordingSegmentStorageMeta]:
+    """Return direct-storage metadata derived from our Clickhouse table."""
+    if not has_archived_segment(project_id, replay_id):
+        return _fetch_segments_from_snuba(project_id, replay_id, offset, limit)
+    return []
+
+
+def fetch_direct_storage_segment_meta(
+    project_id: int,
+    replay_id: str,
+    segment_id: int,
+) -> RecordingSegmentStorageMeta | None:
+    """Return direct-storage metadata derived from our Clickhouse table."""
+    if has_archived_segment(project_id, replay_id):
+        return None
+
+    results = _fetch_segments_from_snuba(
+        project_id=project_id,
+        replay_id=replay_id,
+        offset=0,
+        limit=1,
+        segment_id=segment_id,
+    )
+    if len(results) == 0:
+        return None
+    else:
+        return results[0]
+
+
+def has_archived_segment(project_id: int, replay_id: str) -> bool:
+    """Return true if an archive row exists for this replay."""
+    snuba_request = Request(
+        dataset="replays",
+        app_id="replay-backend-web",
+        query=Query(
+            match=Entity("replays"),
+            select=[Column("is_archived")],
+            where=[
+                Condition(Column("is_archived"), Op.EQ, 1),
+                Condition(Column("project_id"), Op.EQ, project_id),
+                Condition(Column("replay_id"), Op.EQ, str(uuid.UUID(replay_id))),
+                # We request the full 90 day range. This is effectively an unbounded timestamp
+                # range.
+                Condition(Column("timestamp"), Op.LT, datetime.now()),
+                Condition(Column("timestamp"), Op.GTE, datetime.now() - timedelta(days=90)),
+            ],
+            granularity=Granularity(3600),
+        ),
+    )
+    response = raw_snql_query(snuba_request, "replays.query.download_replay_segments")
+    return len(response["data"]) > 0
+
+
+def _fetch_segments_from_snuba(
+    project_id: int,
+    replay_id: str,
+    offset: int,
+    limit: int,
+    segment_id: int | None = None,
+) -> List[RecordingSegmentStorageMeta]:
+    conditions = []
+    if segment_id:
+        conditions.append(Condition(Column("segment_id"), Op.EQ, segment_id))
+
+    snuba_request = Request(
```

Comment:
would it be possible to check the performance of this query given a large dataset? could start think about adding a btree index for replay_id sooner if this is slow (say greater than 1s)

Reply:
Its like 250ms.  Its a "full" table scan.  Its intelligent enough to skip certain projects and replays (which is fast) but it does have to scan from day 0 to day 90.